lib/spack/llnl/util/filesystem.py

```diff
@@ -152,6 +152,31 @@ def lookup(name):
     shutil.copystat = copystat
 
 
+def polite_path(components):
+    """
+    Given a list of strings which are intended to be path components,
+    generate a path, and format each component to avoid generating extra
+    path entries.
+
+    For example all "/", "\", and ":" characters will be replaced with
+    "_". Other characters like "=" will also be replaced.
+    """
+    return os.path.join(*[polite_filename(x) for x in components])
+
+
+def polite_filename(filename: str) -> str:
+    """
+    Replace generally problematic filename characters with underscores.
+
+    This differs from sanitize_filename in that it is more aggressive in
+    changing characters in the name. For example it removes "=" which can
+    confuse path parsing in external tools.
+    """
+    # This character set applies for both Windows and Linux. It does not
+    # account for reserved filenames in Windows.
+    return re.sub(r"[^A-Za-z0-9_.-]", "_", filename)
```

Comment:
Can you move regex compilation out of the function, this is likely a "hot" function since it's called for each component.